### PR TITLE
qf-mvcar: fix rename carrier side effects + lots of other KNM things

### PR DESCRIPTION
**Pull request by fenollp**
4.0 version coming up

### CODE REVIEW

File: core/kazoo_number_manager/src/providers/knm_providers.erl [hunk 1]
```diff
@@ -142,15 +143,18 @@ service_name(Feature) -> Feature.
...
+    Allowed = cleanse_features(list_allowed_features(Parameters)),
+    Denied = cleanse_features(list_denied_features(Parameters)),
     Available = [Feature
                  || Feature <- Allowed,
                     not lists:member(Feature, Denied)
                 ],
     ?LOG_DEBUG("available features: ~s", [?PP(Available)]),
     Available.
 
+cleanse_features(Features) ->
```
Comment (k-anderson) [inline core/kazoo_number_manager/src/providers/knm_providers.erl]: @fenollp missing spec

File: core/kazoo_number_manager/src/providers/knm_providers.erl [hunk 2]
```diff
@@ -251,20 +255,18 @@ used_by_denied_features(#feature_parameters{used_by = <<"trunkstore">>}) -> [];
...
+    ?LOG_DEBUG("denying external features for number used by ~s: ~s", [UsedBy, ?PP(Features)]),
     Features.
 
 -spec number_denied_features(feature_parameters()) -> ne_binaries().
 number_denied_features(#feature_parameters{denied_features = DeniedFeatures}) ->
-    ?LOG_DEBUG("denied number features set on number document: ~s", [?PP(DeniedFeatures)]),
+    ?LOG_DEBUG("denied features set on number document: ~s", [?PP(DeniedFeatures)]),
     DeniedFeatures.
 
 maybe_deny_admin_only_features(#feature_parameters{is_admin = true}) -> [];
```
Comment (k-anderson) [inline core/kazoo_number_manager/src/providers/knm_providers.erl]: @fenollp missing spec

File: core/kazoo_number_manager/src/providers/knm_providers.erl [hunk 3]
```diff
@@ -370,23 +372,45 @@ do_exec(T0=#{todo := Ns}, Action) ->
 
 exec(Number, Action=delete) ->
```
Comment (k-anderson) [inline core/kazoo_number_manager/src/providers/knm_providers.erl]: @fenollp missing spec

File: core/kazoo_number_manager/src/providers/knm_providers.erl [hunk 4]
```diff
@@ -370,23 +372,45 @@ do_exec(T0=#{todo := Ns}, Action) ->
...
+    {NewN, AllowedModules, DeniedModules} = maybe_rename_carrier_and_strip_denied(N),
+    case DeniedModules =:= [] of
+        true -> exec(NewN, Action, AllowedModules);
         false ->
-            ?LOG_DEBUG("denied number features ~s", [?PP(DeniedRequests)]),
+            ?LOG_DEBUG("denied feature providers: ~s", [?PP(DeniedModules)]),
             knm_errors:unauthorized()
     end.
 
+maybe_rename_carrier_and_strip_denied(N) ->
```
Comment (k-anderson) [inline core/kazoo_number_manager/src/providers/knm_providers.erl]: @fenollp missing spec

File: core/kazoo_number_manager/src/providers/knm_providers.erl [hunk 5]
```diff
@@ -370,23 +372,45 @@ do_exec(T0=#{todo := Ns}, Action) ->
...
+        false -> {N, AllowedRequests, DeniedRequests};
+        true ->
+            N1 = exec(N, save, [?PROVIDER_RENAME_CARRIER]),
+            N2 = remove_denied_features(N1),
+            {NewAllowed, NewDenied} = split_requests(N2),
+            ?LOG_DEBUG("allowing feature providers: ~s", [?PP(NewAllowed)]),
+            {N2, NewAllowed, NewDenied}
+    end.
+
+remove_denied_features(N) ->
```
Comment (k-anderson) [inline core/kazoo_number_manager/src/providers/knm_providers.erl]: @fenollp missing spec

File: core/kazoo_number_manager/src/providers/knm_providers.erl [hunk 6]
```diff
@@ -370,23 +372,45 @@ do_exec(T0=#{todo := Ns}, Action) ->
...
+    PN = knm_number:phone_number(N),
+    NewPN = knm_phone_number:remove_denied_features(PN),
+    knm_number:set_phone_number(N, NewPN).
+
+%% @public
+-spec features_denied(knm_phone_number:knm_phone_number()) -> ne_binaries().
+features_denied(PN) ->
+    cleanse_features(list_denied_features(feature_parameters(PN))).
+
 split_requests(Number) ->
```
Comment (k-anderson) [inline core/kazoo_number_manager/src/providers/knm_providers.erl]: @fenollp missing spec

File: core/kazoo_number_manager/src/knm_phone_number.erl [hunk 7]
```diff
@@ -691,23 +693,6 @@ ensure_features_defined(PN=#knm_phone_number{features = undefined}) ->
     PN#knm_phone_number{features = ?DEFAULT_FEATURES};
 ensure_features_defined(PN) -> PN.
```
Comment (k-anderson) [inline core/kazoo_number_manager/src/knm_phone_number.erl]: @fenollp missing spec (can't comment on the start of the function)

File: core/kazoo_number_manager/src/providers/knm_providers.erl (hunk 3 above)
Comment (fenollp) [inline core/kazoo_number_manager/src/providers/knm_providers.erl]: nope

File: core/kazoo_number_manager/src/knm_phone_number.erl (hunk 7 above)
Comment (fenollp) [inline core/kazoo_number_manager/src/knm_phone_number.erl]: done

File: core/kazoo_number_manager/src/providers/knm_providers.erl (hunk 3 above)
Comment (k-anderson) [inline core/kazoo_number_manager/src/providers/knm_providers.erl]: oops :)

File: core/kazoo_number_manager/src/knm_phone_number.erl [hunk 8]
```diff
@@ -1526,7 +1505,58 @@ set_created(PN=#knm_phone_number{created = undefined}, undefined) ->
...
+             ,[kz_util:iolist_join($,, lists:usort([ToRm || [ToRm|_] <- RemoveFromPub]))]),
+    NewPvt = kz_json:prune_keys(RemoveFromPvt, features(PN)),
+    NewPub = kz_json:prune_keys(RemoveFromPub, doc(PN)),
+    Updates = [{fun set_features/2, NewPvt}
+              ,{fun set_doc/2, NewPub}
+              ],
+    {ok, NewPN} = setters(PN, Updates),
+    NewPN.
+
+rm_in_pvt(Feature) ->
```
Comment (k-anderson) [inline core/kazoo_number_manager/src/knm_phone_number.erl]: @fenollp spec, +1 for a more descriptive name ;)

File: core/kazoo_number_manager/src/knm_phone_number.erl [hunk 9]
```diff
@@ -1526,7 +1505,58 @@ set_created(PN=#knm_phone_number{created = undefined}, undefined) ->
...
+    case maps:is_key(Feature, pvt_to_pub()) of
+        false -> [];
+        true -> [[Feature]]
+    end.
+
+rm_in_pub(Feature) ->
```
Comment (k-anderson) [inline core/kazoo_number_manager/src/knm_phone_number.erl]: @fenollp spec, +1 for a more descriptive name ;)

File: core/kazoo_number_manager/src/knm_phone_number.erl [hunk 10]
```diff
@@ -1526,7 +1505,58 @@ set_created(PN=#knm_phone_number{created = undefined}, undefined) ->
...
+    maps:get(Feature, pvt_to_pub(), []).
+
+pvt_to_pub() ->
```
Comment (k-anderson) [inline core/kazoo_number_manager/src/knm_phone_number.erl]: @fenollp spec, +1 for a more descriptive name ;)

**Pull request merged**